src/components/menu/menu.ts

```diff
@@ -256,25 +325,112 @@ export class Menu extends Component<Props, SpreadsheetChildEnv> {
+        this.openSubMenuViaMouse(menu, ev);
       } else {
         this.activateMenu(menu);
       }
     }
   }
 
-  onMouseOver(menu: Action, position: Pixel, ev: MouseEvent) {
+  onMouseMove(menu: Action, ev: MouseEvent) {
+    this.subMenu.shouldSelectFirstItem = false;
+    this.subMenu.menuIndex = this.clickableMenuItems.findIndex(
+      (menuItem) => menuItem.id === menu.id
+    );
     if (this.isEnabled(menu)) {
       if (this.isRoot(menu)) {
-        this.openSubMenu(menu, position, ev);
+        this.openSubMenuViaMouse(menu, ev);
       } else {
         this.closeSubMenu();
       }
     }
   }
+
+  onKeydown(ev: KeyboardEvent) {
+    ev.preventDefault();
+    ev.stopPropagation();
+    if (ev.key.startsWith("Arrow")) {
+      this.processArrows(ev);
+    } else if (ev.key === "Enter") {
+      if (this.subMenu.menuIndex === undefined) return;
+      const menu = this.clickableMenuItems[this.subMenu.menuIndex];
+      if (this.isEnabled(menu)) {
+        if (this.isRoot(menu)) {
+          this.openSubMenuViaKeyboard();
+        } else {
+          this.activateMenu(menu);
+        }
+      }
+    } else if (ev.key === "Escape") {
+      this.close();
+    }
+  }
+
+  processArrows(ev: KeyboardEvent) {
+    switch (ev.key) {
+      case "ArrowDown": {
+        this.selectNextMenuItem();
+        break;
+      }
+      case "ArrowUp": {
+        this.selectLastMenuItem();
+        break;
+      }
+      case "ArrowRight": {
+        this.openSubMenuViaKeyboard();
+        break;
+      }
+      case "ArrowLeft": {
+        if (!this.props.backToParent) return;
+        if (!this.subMenu.isOpen) {
+          this.props.backToParent("left", false);
+          return;
+        }
+        this.closeSubMenu();
+        break;
+      }
+    }
+  }
+
+  forSubMenuToComeBack(direction: string, propagation: boolean) {
+    this.closeSubMenu();
+    if (this.props.backToParent === undefined) return;
+    if (propagation) {
+      this.props.backToParent(direction);
+    }
+  }
+
+  private selectNextMenuItem() {
+    if (this.subMenu.menuIndex === undefined) {
+      this.subMenu.menuIndex = 0;
+    } else if (this.subMenu.menuIndex === this.clickableMenuItems.length - 1) {
+      this.subMenu.menuIndex = 0;
+    } else {
+      this.subMenu.menuIndex++;
+    }
+    const el = this.menuItemRefs[this.subMenu.menuIndex]?.ref.el;
+    el?.scrollIntoView(false);
+  }
+
+  private selectLastMenuItem() {
+    if (this.subMenu.menuIndex === undefined) {
+      this.subMenu.menuIndex = this.clickableMenuItems.length - 1;
+    } else if (this.subMenu.menuIndex === 0) {
+      this.subMenu.menuIndex = this.clickableMenuItems.length - 1;
+    } else {
+      this.subMenu.menuIndex--;
+    }
+    const el = this.menuItemRefs[this.subMenu.menuIndex].ref?.el;
+    el?.scrollIntoView(false);
+  }
+
+  private backToParentRight() {
+    if (!this.props.backToParent) return;
+    this.props.backToParent("right", true);
```

Comment:
you don't have to early return
```suggestion
    this.props?.backToParent("right", true);
```

Reply:
`backToParent` is called here so I cannot use `?` to avoid early return :(

Reply:
Sorry the suggestion was bad, I was more thinkng of something like:
you don't have to early return
```suggestion
    this.props.backToParent?.("right", true);
```